sstables/sstable_set.cc

```diff
@@ -769,6 +770,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+            auto rs = _reader_queue.pop(next_peeked_pos, timeout);
+            for (auto& r: rs) {
+                _all_readers.push_front(std::move(r));
+                _unpeeked_readers.push_back(_all_readers.begin());
+            }
+            return peek_readers(timeout).then([this, timeout] { return (*this)(timeout); });
+        }
+
+        if (_peeked_readers.empty()) {
+            // We are either in forwarding mode and waiting for a fast-forward,
+            // or we've exhausted all the readers.
+            _current_batch.clear();
+            if (_should_return_partition_end) {
+                // Not forwarding, so all readers must be exhausted. Return the last fragment.
+                _current_batch.push_back(mutation_fragment(*_schema, _permit, partition_end()));
+                _should_return_partition_end = false;
+            }
+            return make_ready_future<mutation_fragment_batch>(_current_batch);
+        }
+
+        // Take all fragments with the next smallest position (there may be multiple such fragments).
+        _current_batch.clear();
+        do {
+            boost::range::pop_heap(_peeked_readers, _peeked_cmp);
+            auto r = _peeked_readers.back();
+            auto mf = r->reader.pop_mutation_fragment();
+            _peeked_readers.pop_back();
+            _unpeeked_readers.push_back(std::move(r));
+            _current_batch.push_back(std::move(mf));
+        } while (!_peeked_readers.empty()
+                    && _cmp(_current_batch.back().position(), _peeked_readers.front()->reader.peek_buffer().position()) == 0);
+
+        if (_unpeeked_readers.size() == 1 && _unpeeked_readers.front() == _galloping_reader) {
+            // The first condition says that only one reader was moved from the heap,
+            // i.e. all other readers had strictly greater positions.
+            // The second condition says that this reader already was a galloping candidate,
+            // so let's increase his score.
+            ++_gallop_mode_hits;
+
+            if (in_gallop_mode()) {
+                // We've entered gallop mode with _galloping_reader.
+                // In the next operator() call we will peek this reader on a separate codepath,
+                // using _galloping_reader instead of _unpeeked_readers.
+                _unpeeked_readers.clear();
+            }
+        } else {
+            // Each reader currently in _unpeeked_readers is a potential galloping candidate
+            // (they won with all other readers in _peeked_readers). Remember one of them.
+            _galloping_reader = _unpeeked_readers.front();
+            _gallop_mode_hits = 1;
+        }
+
+
+        collect_exhausted_readers(_current_batch.back().position());
+
+        return make_ready_future<mutation_fragment_batch>(_current_batch);
+    }
+
+    void next_partition() {
+        on_internal_error(cormlogger, "next_partition: this reader works only for single partition queries");
```

Comment:
We usually just throw `std::bad_function_call` from such un-implemented methods.

Reply:
According to the docs of `bad_function_call`, it seems completely relevant to this problem:
https://en.cppreference.com/w/cpp/utility/functional/bad_function_call

"std::bad_function_call is the type of the exception thrown by std::function::operator() if the function wrapper has no target. "

It also doesn't support including an error message :(

I can throw `std::runtime_error` with this error message if we don't want `on_internal_error` here (although it is a bug if someone uses this reader for non-single-partition queries)

Reply:
I think there is value in being consistent. We should handle unimplemented reader methods the same way everywhere. Instead of adding a message, we throw with a backtrace in other places, see https://github.com/scylladb/scylla/blob/052a8d036d8c5ef73cb2ace333def29accde3634/mutation_reader.cc#L1067-L1069